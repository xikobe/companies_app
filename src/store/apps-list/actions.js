import axios from 'axios';
import {
  FETCH_APPS_PENDING,
  FETCH_APPS_REJECTED,
  FETCH_APPS_FULFILLED,
} from './constants';

export const fetchApps = () => async (dispatch) => {
  dispatch({ type: FETCH_APPS_PENDING });

  try {
    const { data } = await axios.get('apps.json');

    dispatch({ type: FETCH_APPS_FULFILLED, payload: data });
  } catch (error) {
    dispatch({ type: FETCH_APPS_REJECTED, payload: error });
  }
};
