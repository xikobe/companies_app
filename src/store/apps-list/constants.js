export const FETCH_APPS_PENDING = 'FETCH_APPS_PENDING';
export const FETCH_APPS_REJECTED = 'FETCH_APPS_REJECTED';
export const FETCH_APPS_FULFILLED = 'FETCH_APPS_FULFILLED';
