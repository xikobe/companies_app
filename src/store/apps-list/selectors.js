import get from 'lodash/get';

export const selectAppsList = (state) => state.appsList;

export const selectAppsListData = (state) => get(selectAppsList(state), 'data');
export const selectAppsListIsLoading = (state) => get(selectAppsList(state), 'isLoading');
export const selectAppsListIsError = (state) => get(selectAppsList(state), 'error');
export const selectAppsCategories = (state) => get(selectAppsList(state), 'categories');
export const selectAppsCategoriesList = (state) => get(selectAppsCategories(state), 'list', []);
