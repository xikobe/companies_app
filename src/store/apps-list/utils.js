/*
 * Creates a normalized object with apps data (including search string)
 * and an array of ids.
 */
export const parseAppsData = (appsList) => appsList.reduce((result, app) => ({
  apps: {
    ...result.apps,
    [app.id]: {
      ...app,
      searchTerms: `
        ${app.name},
        ${app.description},
        ${app.categories},
        ${app.subscriptions.map((sub) => sub.name)}
        `,
    },
  },
  ids: [...result.ids, app.id],
}), {
  apps: {}, ids: [], categoryList: [], categoryApps: {},
});

/*
 * Parses categories into a normalized object of categories containing ids
 * for corresponding apps. And a sorted array of the available categories.
 */
export const parseCategories = (appsList) => appsList.reduce((result, app) => {
  const byCategory = app.categories.reduce((cats, cat) => ({
    ...cats,
    [cat]: [...(result.appsByCategory[cat] ? result.appsByCategory[cat] : []), app.id],
  }), {});

  return {
    ...result,
    list: [
      ...result.list,
      ...app.categories.filter((category) => !result.list.includes(category)),
    ].sort(),
    appsByCategory: {
      ...result.appsByCategory,
      ...byCategory,
    },
  };
}, { list: [], appsByCategory: {} });
