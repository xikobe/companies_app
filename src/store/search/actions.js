import { SEARCH_CHANGE, RESET_SEARCH } from './constants';

export const setSearch = (value) => ({
  type: SEARCH_CHANGE,
  payload: value,
});

export const resetSearch = () => ({
  type: RESET_SEARCH,
});
