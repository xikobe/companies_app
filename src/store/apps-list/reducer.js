import { combineReducers } from 'redux';
import { FETCH_APPS_PENDING, FETCH_APPS_FULFILLED, FETCH_APPS_REJECTED } from './constants';
import { parseCategories, parseAppsData } from './utils';

const initialState = {
  data: {},
  isLoading: false,
  error: null,
  categories: {},
};

const data = (state = initialState.data, { type, payload } = {}) => {
  switch (type) {
    case FETCH_APPS_FULFILLED:
      return parseAppsData(payload);
    default:
      return state;
  }
};

const error = (state = initialState.error, { type, payload } = {}) => {
  switch (type) {
    case FETCH_APPS_REJECTED:
      return payload;
    default:
      return state;
  }
};

const isLoading = (state = initialState.isLoading, { type } = {}) => {
  switch (type) {
    case FETCH_APPS_PENDING:
      return true;
    case FETCH_APPS_FULFILLED:
    case FETCH_APPS_REJECTED:
      return false;
    default:
      return state;
  }
};

export const categories = (state = initialState.categories, { type, payload }) => {
  switch (type) {
    case FETCH_APPS_FULFILLED:
      return parseCategories(payload);
    default:
      return state;
  }
};

export default combineReducers({
  data,
  isLoading,
  error,
  categories,
});
