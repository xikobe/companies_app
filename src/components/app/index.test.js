import React from 'react';
import { shallow, mount } from 'enzyme';
import { useDispatch } from 'react-redux';
import App from './index';
import { fetchApps } from '../../store/apps-list/actions';

jest.mock('react-redux');
jest.mock('./styles');
jest.mock('../../store/apps-list/actions');

const mockDispatch = () => 'Foo';

useDispatch.mockImplementation(() => mockDispatch);

describe('App', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<App />);

    expect(wrapper).toMatchSnapshot();
  });

  it('fetches apps list', () => {
    const wrapper = mount(<App />);

    expect(fetchApps).toHaveBeenCalled();
  });
});
