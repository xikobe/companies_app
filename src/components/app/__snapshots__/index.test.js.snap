// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<Styled(main)>
  <CategoriesProvider>
    <Categories />
    <Styled(main)>
      <Search />
      <List />
    </Styled(main)>
  </CategoriesProvider>
</Styled(main)>
`;
